File: janitor/debdiff.py
#!/usr/bin/python3
# Copyright (C) 2019 Jelmer Vernooij <jelmer@jelmer.uk>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA

import asyncio
import re
from typing import Iterator, Tuple, List, Optional


def iter_sections(text: str) -> Iterator[Tuple[Optional[str], List[str]]]:
    lines = list(text.splitlines(False))
    title = None
    paragraph: List[str] = []
    i = 0
    while i < len(lines):
        line = lines[i]
        if i+1 < len(lines) and lines[i+1] == (len(line) * '-'):
            if title or paragraph:
                yield title, paragraph
            title = line
            paragraph = []
            i += 1
        elif not line.rstrip('\n'):
            if title or paragraph:
                yield title, paragraph
            title = None
            paragraph = []
        else:
            paragraph.append(line)
        i += 1
    if title or paragraph:
        yield title, paragraph


def filter_boring_wdiff(
        lines: List[str], old_version: str, new_version: str) -> Optional[str]:
    if not lines:
        return lines
    field, changes = lines[0].split(':', 1)
    if field == 'Installed-Size':
        return []
    if field == 'Version':
        return []
    lines = [re.sub(
        r'\[-%s(.*?)-\] \{\+%s\1\+\}' % (
            re.escape(old_version), re.escape(new_version)),
        '', line) for line in lines]
    block = '\n'.join(lines)
    if (not re.findall(r'\[-.*?-\]', block) and
            not re.findall('\{\+.*?\+\}', block)):
        return []
    return lines


def _iter_fields(lines):
    cl = []
    for line in lines:
        if cl and line.startswith(' '):
            cl.append(line)
        else:
            yield cl
            cl = [line]
    if cl:
        yield cl


def filter_boring(debdiff: str, old_version: str, new_version: str) -> str:
    ret: List[Tuple[Optional[str], List[str]]] = []
    for title, paragraph in iter_sections(debdiff):
        if not title:
            ret.append((title, paragraph))
            continue
        package: Optional[str]
        m = re.match(
                r'Control files of package (.*): lines which differ '
                r'\(wdiff format\)',
                title)
        if m:
            package = m.group(1)
            wdiff = True
        elif title == 'Control files: lines which differ (wdiff format)':
            package = None
            wdiff = True
        else:
            package = None
            wdiff = False
        if wdiff:
            paragraph_unfiltered = []
            for lines in _iter_fields(paragraph):
                newlines = filter_boring_wdiff(lines, old_version, new_version)
                paragraph_unfiltered.extend(newlines)
            paragraph = [line for line in paragraph_unfiltered
                         if line is not None]
            if any([line.strip() for line in paragraph]):
                ret.append((title, paragraph))
            else:
                if package:
                    ret.append((
                        None,
                        ['No differences were encountered between the control '
                         'files of package %s' % package]))
                else:
                    ret.append((
                        None,
                        ['No differences were encountered in the control files'
                         ]))
        else:
            ret.append((title, paragraph))

    lines = []
    for title, paragraph in ret:
        if title is not None:
            lines.append(title)
            lines.append(len(title) * '-')
        lines.extend(paragraph)
        lines.append('')
    return '\n'.join(lines)


class DebdiffError(Exception):
    """Error occurred while running debdiff."""


async def run_debdiff(
        old_binaries: List[str], new_binaries: List[str]) -> bytes:
    args = (['debdiff', '--from'] +
            old_binaries + ['--to'] + new_binaries)
    p = await asyncio.create_subprocess_exec(
        *args, stdin=asyncio.subprocess.PIPE,
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE)
    stdout, stderr = await p.communicate(b'')
    if p.returncode not in (0, 1):
        raise DebdiffError(stderr.decode(errors='replace'))
    return stdout


def debdiff_is_empty(debdiff: str) -> bool:
    return any(
        [title is not None for (title, paragraph) in iter_sections(debdiff)])


def section_is_wdiff(title: str) -> Tuple[bool, Optional[str]]:
    m = re.match(
            r'Control files of package (.*): lines which differ '
            r'\(wdiff format\)',
            title)
    if m:
        return (True, m.group(1))
    if title == 'Control files: lines which differ (wdiff format)':
        return (True, None)
    return (False, None)


def markdownify_debdiff(debdiff: str) -> str:
    def fix_wdiff_md(line):
        # GitLab markdown will render links but then not show the
        # delta highlighting. This fools it into not autolinking:
        return line.replace('://', '&#8203;://')
    ret = []
    for title, lines in iter_sections(debdiff):
        if title:
            ret.append("### %s" % title)
            wdiff, package = section_is_wdiff(title)
            if wdiff:
                ret.extend(
                    ["* %s" % fix_wdiff_md(line)
                     for line in lines if line.strip()])
            else:
                for line in lines:
                    ret.append('    ' + line)
        else:
            ret.append("")
            for line in lines:
                if line.strip():
                    line = re.sub(
                        '^(No differences were encountered between the '
                        'control files of package) (.*)$',
                        r'\1 \*\*\2\*\*', line)
                    ret.append(line)
                else:
                    ret.append("")
            if ret[-1] == "":
                ret.pop(-1)
    return "\n".join(ret)


def htmlize_debdiff(debdiff: str) -> str:
    def highlight_wdiff(line):
        line = re.sub(
            r'\[-(.*?)-\]',
            r'<span style="color:red;font-weight:bold">\1</span>', line)
        line = re.sub(
            r'\{\+(.*?)\+\}',
            r'<span style="color:green;font-weight:bold">\1</span>', line)
        return line
    ret = []
    for title, lines in iter_sections(debdiff):
        if title:
            ret.append("<h4>%s</h4>" % title)
            if re.match(
                    r'Control files of package .*: lines which differ '
                    r'\(wdiff format\)',
                    title):
                wdiff = True
            elif title == 'Control files: lines which differ (wdiff format)':
                wdiff = True
            else:
                wdiff = False
            if wdiff:
                ret.append("<ul>")
                for line in _iter_fields(lines):
                    if not line:
                        continue
                    ret.extend(["<li><pre>%s</pre></li>" % highlight_wdiff('\n'.join(line))])
                ret.append("</ul>")
            else:
                ret.append("<pre>")
                ret.extend(lines)
                ret.append("</pre>")
        else:
            ret.append("<p>")
            for line in lines:
                if line.strip():
                    line = re.sub(
                        '^(No differences were encountered between the '
                        'control files of package) (.*)$',
                        '\\1 <b>\\2</b>', line)
                    ret.append(line)
                else:
                    ret.append("</p>")
                    ret.append("<p>")
            if ret[-1] == "<p>":
                ret.pop(-1)
    return "\n".join(ret)

File: src/debdiff.rs
/// Parse a debdiff
///
/// # Arguments
/// * `text` - The debdiff
///
/// # Returns
/// An iterator over the sections of the debdiff
/// Each section is a tuple of the title of the section and the lines of the section
/// The title is None if the section is a paragraph
/// The lines are a Vec of strings
pub fn iter_sections(text: &str) -> impl Iterator<Item = (Option<&str>, Vec<&str>)> {
    let lines = text.split_terminator('\n').collect::<Vec<_>>();
    let mut title = None;
    let mut paragraph = Vec::new();
    let mut i = 0;
    let mut ret = vec![];
    while i < lines.len() {
        let line = lines[i];
        if i + 1 < lines.len() && lines[i + 1] == ("-".repeat(line.len())) {
            if title.is_some() || !paragraph.is_empty() {
                ret.push((title, paragraph));
            }
            title = Some(line);
            paragraph = Vec::new();
            i += 1;
        } else if line.trim_end().is_empty() {
            if title.is_some() || !paragraph.is_empty() {
                ret.push((title, paragraph));
            }
            title = None;
            paragraph = Vec::new();
        } else {
            paragraph.push(line);
        }
        i += 1;
    }
    if title.is_some() || !paragraph.is_empty() {
        ret.push((title, paragraph));
    }
    ret.into_iter()
}

/// Filter boring lines out of a wdiff
///
/// # Arguments
/// * `lines` - The lines of the wdiff
/// * `old_version` - The old version
/// * `new_version` - The new version
///
/// # Returns
/// A new Vec of lines with the boring lines removed
pub fn filter_boring_wdiff<'a>(
    lines: Vec<&'a str>,
    old_version: &'a str,
    new_version: &'a str,
) -> Vec<String> {
    if lines.is_empty() {
        return Vec::new();
    }
    let (field, _changes) = match lines[0].split_once(':') {
        Some((field, changes)) => (field, changes),
        None => {
            return lines
                .iter()
                .map(|line| line.to_string())
                .collect::<Vec<_>>()
        }
    };
    if field == "Installed-Size" {
        return Vec::new();
    }
    if field == "Version" {
        return Vec::new();
    }
    let regex = fancy_regex::Regex::new(&format!(
        r"\[-{}(.*?)-\] \{{\+{}\1\+\}}",
        regex::escape(old_version),
        regex::escape(new_version),
    ))
    .unwrap();

    let lines = lines
        .iter()
        .map(|line| regex.replace_all(line, r"$1").to_string())
        .collect::<Vec<_>>();
    let block = lines.join("\n");

    if lazy_regex::regex_find!(r"\[-.*?-\]", &block).is_none()
        && lazy_regex::regex_find!(r"\{\+.*?\+\}", &block).is_none()
    {
        return Vec::new();
    }
    lines
}

fn iter_fields<'a>(
    lines: impl Iterator<Item = &'a str> + 'a,
) -> impl Iterator<Item = Vec<&'a str>> + 'a {
    let mut cl = Vec::new();
    let mut ret = Vec::new();
    for line in lines {
        if !cl.is_empty() && line.starts_with(" ") {
            cl.push(line);
        } else {
            if !cl.is_empty() {
                ret.push(cl);
            }
            cl = vec![line];
        }
    }
    if !cl.is_empty() {
        ret.push(cl);
    }
    ret.into_iter()
}

/// Filter boring lines out of a debdiff
///
/// # Arguments
/// * `debdiff` - The debdiff
/// * `old_version` - The old version
/// * `new_version` - The new version
///
/// # Returns
/// A new debdiff with the boring lines removed
pub fn filter_boring(debdiff: &str, old_version: &str, new_version: &str) -> String {
    let mut ret = Vec::new();
    for (title, paragraph) in iter_sections(debdiff) {
        let title = if let Some(title) = title {
            title
        } else {
            ret.push((
                title,
                paragraph
                    .iter()
                    .map(|line| line.to_string())
                    .collect::<Vec<_>>(),
            ));
            continue;
        };
        let (package, wdiff) = if let Some((_, package)) = lazy_regex::regex_captures!(
            r"Control files of package (.*): lines which differ \(wdiff format\)",
            title
        ) {
            (Some(package), true)
        } else if title == "Control files: lines which differ (wdiff format)" {
            (None, true)
        } else {
            (None, false)
        };
        if wdiff {
            let mut paragraph_unfiltered = Vec::new();
            for lines in iter_fields(paragraph.into_iter()) {
                let newlines = filter_boring_wdiff(lines, old_version, new_version);
                paragraph_unfiltered.extend(newlines);
            }
            let paragraph = paragraph_unfiltered;
            if paragraph.iter().any(|line| !line.trim().is_empty()) {
                ret.push((Some(title), paragraph));
            } else if let Some(package) = package {
                ret.push((
                    None,
                    vec![format!(
                        "No differences were encountered between the control files of package {}",
                        package
                    )],
                ));
            } else {
                ret.push((
                    None,
                    vec!["No differences were encountered in the control files".to_string()],
                ));
            }
        } else {
            ret.push((
                Some(title),
                paragraph
                    .iter()
                    .map(|line| line.to_string())
                    .collect::<Vec<_>>(),
            ));
        }
    }

    let mut lines = vec![];
    for (title, paragraph) in ret {
        if let Some(title) = title {
            lines.push(title.to_string());
            lines.push("-".repeat(title.len()));
        }
        lines.extend(paragraph);
        lines.push("".to_string());
    }
    lines.join("\n")
}

#[derive(Debug)]
pub struct DebdiffError {
    message: String,
}

impl From<tokio::io::Error> for DebdiffError {
    fn from(e: tokio::io::Error) -> Self {
        DebdiffError {
            message: e.to_string(),
        }
    }
}

impl std::fmt::Display for DebdiffError {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "DebdiffError: {}", self.message)
    }
}

impl std::error::Error for DebdiffError {}

pub async fn run_debdiff(
    old_binaries: Vec<&str>,
    new_binaries: Vec<&str>,
) -> Result<Vec<u8>, DebdiffError> {
    // Validate all file paths to prevent command injection
    for path in old_binaries.iter().chain(new_binaries.iter()) {
        // Check for shell metacharacters and path traversal
        if path.contains("..") || 
           path.chars().any(|c| matches!(c, ';' | '&' | '|' | '$' | '`' | '\n' | '\r' | '\0')) {
            return Err(DebdiffError {
                message: format!("Invalid file path: {}", path),
            });
        }
        
        // Verify the path exists and is a regular file
        let path_obj = std::path::Path::new(path);
        if !path_obj.exists() {
            return Err(DebdiffError {
                message: format!("File not found: {}", path),
            });
        }
        if !path_obj.is_file() {
            return Err(DebdiffError {
                message: format!("Not a regular file: {}", path),
            });
        }
    }
    
    let mut cmd = tokio::process::Command::new("debdiff");
    cmd.arg("--from");
    for path in old_binaries {
        cmd.arg(path);
    }
    cmd.arg("--to");
    for path in new_binaries {
        cmd.arg(path);
    }
    
    let output = cmd.output().await?;
    if !output.status.success() {
        return Err(DebdiffError {
            message: String::from_utf8_lossy(&output.stderr).to_string(),
        });
    }
    Ok(output.stdout)
}

pub fn debdiff_is_empty(debdiff: &str) -> bool {
    !iter_sections(debdiff).any(|(title, _paragraph)| title.is_some())
}

pub fn section_is_wdiff(title: &str) -> (bool, Option<&str>) {
    if let Some((_, package)) = lazy_regex::regex_captures!(
        r"Control files of package (.*): lines which differ \(wdiff format\)",
        title
    ) {
        return (true, Some(package));
    }
    if title == "Control files: lines which differ (wdiff format)" {
        return (true, None);
    }
    (false, None)
}

pub fn markdownify_debdiff(debdiff: &str) -> String {
    let fix_wdiff_md = |line: &str| {
        // GitLab markdown will render links but then not show the
        // delta highlighting. This fools it into not autolinking:
        line.replace("://", "&#8203;://")
    };

    let mut ret = vec![];
    for (title, lines) in iter_sections(debdiff) {
        if let Some(title) = title {
            ret.push(format!("### {}", title));
            let (wdiff, _package) = section_is_wdiff(title);
            if wdiff {
                ret.extend(lines.iter().filter_map(|line| {
                    if line.trim().is_empty() {
                        None
                    } else {
                        Some(format!("* {}", fix_wdiff_md(line)))
                    }
                }));
            } else {
                for line in lines {
                    ret.push(format!("    {}", line));
                }
            }
        } else {
            ret.push("".to_owned());
            for line in lines {
                if !line.trim().is_empty() {
                    let line = lazy_regex::regex_replace!(
                        "^(No differences were encountered between the control files of package) (.*)$",
                        r"$1 \*\*$2\*\*",
                        line,
                    );
                    ret.push(line.to_string());
                } else {
                    ret.push("".to_owned());
                }
            }
            if ret.last() == Some(&String::new()) {
                ret.pop();
            }
        }
    }
    ret.join("\n")
}

pub fn htmlize_debdiff(debdiff: &str) -> String {
    let highlight_wdiff = |line| {
        let line = lazy_regex::regex_replace!(
            r"\[-(.*?)-\]",
            r#"<span style="color:red;font-weight:bold">$1</span>"#,
            line
        );
        let line = lazy_regex::regex_replace!(
            r"\{\+(.*?)\+\}",
            r#"<span style="color:green;font-weight:bold">$1</span>"#,
            line,
        );
        line
    };

    let mut ret = vec![];
    for (title, lines) in iter_sections(debdiff) {
        if let Some(title) = title {
            ret.push(format!("<h4>{}</h4>", title));
            let wdiff = if lazy_regex::regex_is_match!(
                r"Control files of package .*: lines which differ \(wdiff format\)",
                title,
            ) {
                true
            } else {
                title == "Control files: lines which differ (wdiff format)"
            };
            if wdiff {
                ret.push("<ul>".to_owned());
                for mlines in iter_fields(lines.into_iter()) {
                    if mlines.is_empty() {
                        continue;
                    }
                    ret.push(format!(
                        "<li><pre>{}</pre></li>",
                        highlight_wdiff(mlines.join("\n"))
                    ));
                }
                ret.push("</ul>".to_owned());
            } else {
                ret.push("<pre>".to_owned());
                ret.extend(
                    lines
                        .iter()
                        .map(|line| line.to_string())
                        .collect::<Vec<_>>(),
                );
                ret.push("</pre>".to_owned());
            }
        } else {
            ret.push("<p>".to_owned());
            for line in lines {
                if !line.trim().is_empty() {
                    let line = lazy_regex::regex_replace!(
                        "^(No differences were encountered between the control files of package) (.*)$",
                        r"$1 <b>$2</b>",
                        line,
                    ).to_string();
                    ret.push(line);
                } else {
                    ret.push("</p>".to_owned());
                    ret.push("<p>".to_owned());
                }
            }
            if ret.last().unwrap() == "<p>" {
                ret.pop();
            }
        }
    }
    ret.join("\n")
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn test_nothing() {
        assert_eq!(
            iter_sections("foo\n").collect::<Vec<_>>(),
            vec![(None, vec!["foo"])]
        );
    }

    #[test]
    fn test_simple() {
        assert_eq!(
        vec![
            (
                None,
                vec![
                    "[The following lists of changes regard files as different if they have",
                    "different names, permissions or owners.]",
                ],
            ),
            (
                Some("Files in second .changes but not in first"),
                vec!["-rw-r--r--  root/root   /usr/lib/debug/.build-id/e4/3520e0f1e.debug"],
            ),
            (
                Some("Files in first .changes but not in second"),
                vec!["-rw-r--r--  root/root   /usr/lib/debug/.build-id/28/0303571bd.debug"],
            ),
            (
                Some("Control files of package xserver-blah: lines which differ (wdiff format)"),
                vec![
                    "Installed-Size: [-174-] {+170+}",
                    "Version: [-1:1.7.9-2~jan+unchanged1-] {+1:1.7.9-3~jan+lint1+}",
                ],
            ),
            (
                Some("Control files of package xserver-dbgsym: lines which differ (wdiff format)"),
                vec![
                    "Build-Ids: [-280303571bd7f8-] {+e43520e0f1eb+}",
                    "Depends: xserver-blah (= [-1:1.7.9-2~jan+unchanged1)-] {+1:1.7.9-3~jan+lint1)+}",
                    "Installed-Size: [-515-] {+204+}",
                    "Version: [-1:1.7.9-2~jan+unchanged1-] {+1:1.7.9-3~jan+lint1+}",
                ],
            ),
        ], iter_sections(r#"[The following lists of changes regard files as different if they have
different names, permissions or owners.]

Files in second .changes but not in first
-----------------------------------------
-rw-r--r--  root/root   /usr/lib/debug/.build-id/e4/3520e0f1e.debug

Files in first .changes but not in second
-----------------------------------------
-rw-r--r--  root/root   /usr/lib/debug/.build-id/28/0303571bd.debug

Control files of package xserver-blah: lines which differ (wdiff format)
------------------------------------------------------------------------
Installed-Size: [-174-] {+170+}
Version: [-1:1.7.9-2~jan+unchanged1-] {+1:1.7.9-3~jan+lint1+}

Control files of package xserver-dbgsym: lines which differ (wdiff format)
--------------------------------------------------------------------------
Build-Ids: [-280303571bd7f8-] {+e43520e0f1eb+}
Depends: xserver-blah (= [-1:1.7.9-2~jan+unchanged1)-] {+1:1.7.9-3~jan+lint1)+}
Installed-Size: [-515-] {+204+}
Version: [-1:1.7.9-2~jan+unchanged1-] {+1:1.7.9-3~jan+lint1+}
"#).collect::<Vec<_>>());
    }

    #[test]
    fn test_just_versions() {
        let debdiff = r#"File lists identical (after any substitutions)

Control files of package acpi-fakekey: lines which differ (wdiff format)
------------------------------------------------------------------------
Version: [-0.143-4~jan+unchanged1-] {+0.143-5~jan+lint1+}

Control files of package acpi-fakekey-dbgsym: lines which differ (wdiff format)
-------------------------------------------------------------------------------
Depends: acpi-fakekey (= [-0.143-4~jan+unchanged1)-] {+0.143-5~jan+lint1)+}
Version: [-0.143-4~jan+unchanged1-] {+0.143-5~jan+lint1+}

Control files of package acpi-support: lines which differ (wdiff format)
------------------------------------------------------------------------
Version: [-0.143-4~jan+unchanged1-] {+0.143-5~jan+lint1+}

Control files of package acpi-support-base: lines which differ (wdiff format)
-----------------------------------------------------------------------------
Version: [-0.143-4~jan+unchanged1-] {+0.143-5~jan+lint1+}
"#;
        let newdebdiff =
            super::filter_boring(debdiff, "0.143-4~jan+unchanged1", "0.143-5~jan+lint1");
        assert_eq!(
            newdebdiff,
            r#"File lists identical (after any substitutions)

No differences were encountered between the control files of package acpi-fakekey

No differences were encountered between the control files of package acpi-fakekey-dbgsym

No differences were encountered between the control files of package acpi-support

No differences were encountered between the control files of package acpi-support-base
"#
        );
    }

    #[test]
    fn test_iter_sections() {
        let debdiff = r#"[The following lists of changes regard files as different if they have
different names, permissions or owners.]

Files in second set of .debs but not in first
---------------------------------------------
-rw-r--r--  root/root   /usr/lib/debug/.build-id/6b/b4fbad299b8e4159451a97bbbb3a3367c6084f.debug
-rw-r--r--  root/root   /usr/lib/debug/.build-id/c3/83853da545bcc080058e063998c4c887b8a54d.debug

Files in first set of .debs but not in second
---------------------------------------------
-rw-r--r--  root/root   /usr/lib/debug/.build-id/b6/c575e21b54becc213ef29458a11cdb7f12cbd6.debug
-rw-r--r--  root/root   /usr/lib/debug/.build-id/f4/749f468a9f6a13812364388d0091ea6a55e951.debug

Control files of package plasma-thunderbolt: lines which differ (wdiff format)
------------------------------------------------------------------------------
Version: [-5.25.4-1~jan+unchanged1-] {+5.25.4-2~jan+lint1+}

Control files of package plasma-thunderbolt-dbgsym: lines which differ (wdiff format)
-------------------------------------------------------------------------------------
Build-Ids: 3b45f929fe8e39e2c1dcf5832eca81b2c2dd3cf3 [-b6c575e21b54becc213ef29458a11cdb7f12cbd6 f4749f468a9f6a13812364388d0091ea6a55e951-] {+6bb4fbad299b8e4159451a97bbbb3a3367c6084f c383853da545bcc080058e063998c4c887b8a54d+}
Depends: plasma-thunderbolt (= [-5.25.4-1~jan+unchanged1)-] {+5.25.4-2~jan+lint1)+}
Version: [-5.25.4-1~jan+unchanged1-] {+5.25.4-2~jan+lint1+}
"#;
        assert_eq!(iter_sections(debdiff).collect::<Vec<_>>(), vec![
            (
                None,
                vec![
                    "[The following lists of changes regard files as different if they have",
                    "different names, permissions or owners.]",
                ],
            ),
            (
                Some("Files in second set of .debs but not in first"),
                vec![
                    "-rw-r--r--  root/root   /usr/lib/debug/.build-id/6b/b4fbad299b8e4159451a97bbbb3a3367c6084f.debug",
                    "-rw-r--r--  root/root   /usr/lib/debug/.build-id/c3/83853da545bcc080058e063998c4c887b8a54d.debug",
                ],
            ),
            (
                Some("Files in first set of .debs but not in second"),
                vec![
                    "-rw-r--r--  root/root   /usr/lib/debug/.build-id/b6/c575e21b54becc213ef29458a11cdb7f12cbd6.debug",
                    "-rw-r--r--  root/root   /usr/lib/debug/.build-id/f4/749f468a9f6a13812364388d0091ea6a55e951.debug",
                ],
            ),
            (
                Some("Control files of package plasma-thunderbolt: lines which differ (wdiff format)"),
                vec![
                    "Version: [-5.25.4-1~jan+unchanged1-] {+5.25.4-2~jan+lint1+}",
                ],
            ),
            (
                Some("Control files of package plasma-thunderbolt-dbgsym: lines which differ (wdiff format)"),
                vec![
                    "Build-Ids: 3b45f929fe8e39e2c1dcf5832eca81b2c2dd3cf3 [-b6c575e21b54becc213ef29458a11cdb7f12cbd6 f4749f468a9f6a13812364388d0091ea6a55e951-] {+6bb4fbad299b8e4159451a97bbbb3a3367c6084f c383853da545bcc080058e063998c4c887b8a54d+}",
                    "Depends: plasma-thunderbolt (= [-5.25.4-1~jan+unchanged1)-] {+5.25.4-2~jan+lint1)+}",
                    "Version: [-5.25.4-1~jan+unchanged1-] {+5.25.4-2~jan+lint1+}",
                ],
            ),
            ]);
    }

    #[test]
    fn test_filter_wdiff_boring() {
        assert_eq!(
            filter_boring_wdiff(
                vec!["Installed-Size: [-174-] {+170+}"],
                "1:1.7.9-2~jan+unchanged1",
                "1:1.7.9-3~jan+lint1"
            ),
            Vec::<String>::new()
        );

        assert_eq!(
            filter_boring_wdiff(
                vec!["Version: [-1:1.7.9-2~jan+unchanged1-] {+1:1.7.9-3~jan+lint1+}"],
                "1:1.7.9-2~jan+unchanged1",
                "1:1.7.9-3~jan+lint1"
            ),
            Vec::<String>::new()
        );

        assert_eq!(
            filter_boring_wdiff(vec!["Depends: xserver-blah (= [-1:1.7.9-2~jan+unchanged1)-] {+1:1.7.9-3~jan+lint1)+}"], "1:1.7.9-2~jan+unchanged1", "1:1.7.9-3~jan+lint1"),
            Vec::<String>::new()
        );
        assert_eq!(
            filter_boring_wdiff(
                vec!["Build-Ids: [-280303571bd7f8-] {+e43520e0f1eb+}"],
                "1.7.9-2~jan+unchanged1",
                "1.7.9-3~jan+lint1"
            ),
            vec!["Build-Ids: [-280303571bd7f8-] {+e43520e0f1eb+}"]
        );
    }

    #[test]
    fn test_iter_fields() {
        assert_eq!(
            iter_fields(
                vec![
                    "foo", " bar", " baz", "qux", " quux", " corge", "grault", " garply", " waldo",
                    "thud",
                ]
                .into_iter()
            )
            .collect::<Vec<_>>(),
            vec![
                vec!["foo", " bar", " baz"],
                vec!["qux", " quux", " corge"],
                vec!["grault", " garply", " waldo"],
                vec!["thud"],
            ]
        );
    }
}
